File: data/demo-seed.ts
import { recomputeConceptCache, withCachedState } from "../domain/learner-model";
import type { Concept, LearnerSnapshot, LearningEvent, RetrievalOutcome } from "../domain/types";
import { createDemoLearningActivities } from "./demo-learning-activities";

const DAY = 86_400_000;
const iso = (now: number, daysAgo = 0) => new Date(now - daysAgo * DAY).toISOString();

type SeedConcept = {
  id: string;
  name: string;
  examImportance: number;
  difficulty: number;
  estimatedMinutes: number;
  mastery: number;
  lastReviewedDays: number;
  history: RetrievalOutcome[];
};

const MICROECONOMICS_CONCEPTS: SeedConcept[] = [
  { id: "c-supply-demand", name: "Supply & Demand", examImportance: 0.88, difficulty: 0.38, estimatedMinutes: 22, mastery: 0.82, lastReviewedDays: 2, history: ["success", "success", "partial"] },
  { id: "c-elasticity", name: "Elasticity", examImportance: 0.95, difficulty: 0.56, estimatedMinutes: 18, mastery: 0.48, lastReviewedDays: 8, history: ["partial", "failure"] },
  { id: "c-consumer-choice", name: "Consumer Choice", examImportance: 0.68, difficulty: 0.62, estimatedMinutes: 24, mastery: 0.64, lastReviewedDays: 4, history: ["success", "partial"] },
  { id: "c-market-structures", name: "Market Structures", examImportance: 0.86, difficulty: 0.6, estimatedMinutes: 20, mastery: 0.61, lastReviewedDays: 6, history: ["partial", "success"] },
  { id: "c-game-theory", name: "Game Theory", examImportance: 0.25, difficulty: 0.7, estimatedMinutes: 26, mastery: 0.31, lastReviewedDays: 3, history: ["partial"] },
  { id: "c-monetary-policy", name: "Monetary Policy", examImportance: 0.82, difficulty: 0.58, estimatedMinutes: 18, mastery: 0.57, lastReviewedDays: 12, history: ["success", "failure"] },
  { id: "c-fiscal-policy", name: "Fiscal Policy", examImportance: 0.72, difficulty: 0.5, estimatedMinutes: 17, mastery: 0.69, lastReviewedDays: 5, history: ["success", "partial"] },
];

const PROMPTS = {
  "c-supply-demand": ["What happens to equilibrium price when demand rises and supply is unchanged?", "Price rises as buyers compete for unchanged supply, moving the market to a new equilibrium."],
  "c-elasticity": ["Why does demand become more elastic when close substitutes exist?", "Buyers can switch when price rises, so quantity demanded responds more strongly to the price change."],
  "c-consumer-choice": ["What does the budget constraint represent?", "The combinations of goods a consumer can afford at given prices and income."],
  "c-market-structures": ["What feature most clearly separates perfect competition from monopoly?", "The degree of market power: competitive firms are price takers, while a monopolist can influence price."],
  "c-game-theory": ["What is a dominant strategy?", "A strategy that gives a player the best outcome regardless of what the other player chooses."],
  "c-monetary-policy": ["How can a higher policy interest rate reduce inflationary pressure?", "It raises borrowing costs, restrains demand and investment, and can reduce upward pressure on prices."],
  "c-fiscal-policy": ["Name one expansionary fiscal-policy action.", "Increasing government spending or reducing taxes to raise aggregate demand."],
} as const;

export function createDemoSnapshot(nowMs = Date.now()): LearnerSnapshot {
  const now = iso(nowMs);
  const userId = "user-amina";
  const courseId = "course-microeconomics";
  const examId = "exam-microeconomics-final";
  const events: LearningEvent[] = [];

  const concepts = MICROECONOMICS_CONCEPTS.map((item, conceptIndex) => {
    events.push({
      id: `evt-seed-${item.id}`, userId, conceptId: item.id, sessionId: null,
      kind: "seed_rating", outcome: item.mastery >= 0.7 ? "success" : item.mastery >= 0.4 ? "partial" : "failure",
      selfRating: null, assistance: "none", responseTimeMs: null, promptId: null, responseText: null,
      masteryBefore: 0, masteryAfter: item.mastery,
      createdAt: iso(nowMs, item.lastReviewedDays + item.history.length + 2),
    });
    item.history.forEach((outcome, index) => events.push({
      id: `evt-history-${item.id}-${index}`, userId, conceptId: item.id, sessionId: "session-prior",
      kind: "retrieval", outcome, selfRating: null, assistance: "none", responseTimeMs: 20_000 + conceptIndex * 1_000,
      promptId: `p-${item.id}`, responseText: null, masteryBefore: item.mastery, masteryAfter: item.mastery,
      createdAt: iso(nowMs, item.lastReviewedDays + item.history.length - index - 1),
    }));
    const base: Concept = {
      id: item.id, courseId, userId, name: item.name, examImportance: item.examImportance,
      difficulty: item.difficulty, estimatedMinutes: item.estimatedMinutes,
      mastery: 0, confidence: 0, predictedRetention: 0, lastReviewedAt: null, nextReviewAt: null,
      retrievalAttempts: 0, successfulRetrievals: 0, failedRetrievals: 0,
      createdAt: iso(nowMs, 40), updatedAt: now,
    };
    return withCachedState(base, recomputeConceptCache(base, events, now));
  });

  return {
    profile: { id: userId, displayName: "Amina", timezone: "Africa/Casablanca", createdAt: iso(nowMs, 40) },
    courses: [{ id: courseId, userId, name: "Microeconomics", createdAt: iso(nowMs, 40) }],
    exams: [{ id: examId, courseId, userId, target: "Microeconomics Final", targetPercent: 85, examDate: iso(nowMs, -9), availableMinutes: 43, isActive: true }],
    concepts,
    relationships: [
      { id: "r-supply-elasticity", fromId: "c-supply-demand", toId: "c-elasticity", kind: "prerequisite" },
      { id: "r-elasticity-markets", fromId: "c-elasticity", toId: "c-market-structures", kind: "prerequisite" },
      { id: "r-choice-markets", fromId: "c-consumer-choice", toId: "c-market-structures", kind: "related" },
      { id: "r-monetary-fiscal", fromId: "c-monetary-policy", toId: "c-fiscal-policy", kind: "related" },
    ],
    prompts: concepts.map((concept) => ({
      id: `p-${concept.id}`,
      conceptId: concept.id,
      promptText: PROMPTS[concept.id as keyof typeof PROMPTS][0],
      modelAnswer: PROMPTS[concept.id as keyof typeof PROMPTS][1],
    })),
    learningActivities: createDemoLearningActivities(),
    events,
    sessions: [],
  };
}

File: ios/Domain/Sources/VellumDomain/DemoSeed.swift
import Foundation

public enum DemoSeed {
    public static func snapshot(now: Date = Date(timeIntervalSince1970: 1_757_030_400)) -> LearnerSnapshot {
        let profile = StudentProfile(id: "user-maya", displayName: "Maya")
        let course = Course(id: "course-micro", name: "Microeconomics", typicalMinutes: 45)
        let exam = Exam(
            id: "exam-final",
            courseId: course.id,
            date: now.addingTimeInterval(11 * LearnerConstants.secondsPerDay),
            targetGrade: 0.85,
            evidenceConfidence: 0.45
        )

        struct Seed {
            var id: String
            var name: String
            var importance: Double
            var difficulty: Double
            var mastery: Double
            var lastSuccessDays: Double
            var extras: [(daysAgo: Double, outcome: RetrievalOutcome)]
        }

        let seeds: [Seed] = [
            .init(id: "c-supply", name: "Supply & Demand", importance: 0.95, difficulty: 0.3, mastery: 0.91, lastSuccessDays: 1, extras: [(1, .success)]),
            .init(id: "c-elasticity", name: "Elasticity", importance: 0.95, difficulty: 0.55, mastery: 0.42, lastSuccessDays: 12, extras: [(12, .failure), (8, .partial)]),
            .init(id: "c-choice", name: "Consumer Choice", importance: 0.72, difficulty: 0.5, mastery: 0.61, lastSuccessDays: 6, extras: [(6, .success)]),
            .init(id: "c-markets", name: "Market Structures", importance: 0.7, difficulty: 0.45, mastery: 0.58, lastSuccessDays: 9, extras: [(9, .partial)]),
            .init(id: "c-fiscal", name: "Fiscal Policy", importance: 0.82, difficulty: 0.4, mastery: 0.68, lastSuccessDays: 5, extras: [(5, .success)]),
            .init(id: "c-monetary", name: "Monetary Policy", importance: 0.84, difficulty: 0.5, mastery: 0.55, lastSuccessDays: 11, extras: [(11, .success), (3, .partial)]),
            .init(id: "c-game", name: "Game Theory", importance: 0.38, difficulty: 0.7, mastery: 0.37, lastSuccessDays: 16, extras: [(16, .failure)]),
        ]

        var events: [LearningEvent] = []
        var concepts: [Concept] = []

        for (index, seed) in seeds.enumerated() {
            if seed.mastery > 0 {
                events.append(
                    LearningEvent(
                        id: "evt-seed-\(seed.id)",
                        conceptId: seed.id,
                        sessionId: nil,
                        kind: .seedRating,
                        outcome: seed.mastery >= 0.75 ? .success : seed.mastery >= 0.4 ? .partial : .failure,
                        promptId: nil,
                        responseText: nil,
                        masteryBefore: 0,
                        masteryAfter: seed.mastery,
                        createdAt: now.addingTimeInterval(-max(seed.lastSuccessDays, 20) * LearnerConstants.secondsPerDay)
                    )
                )
            }
            for (attemptIndex, extra) in seed.extras.enumerated() {
                events.append(
                    LearningEvent(
                        id: "evt-\(seed.id)-\(attemptIndex)",
                        conceptId: seed.id,
                        sessionId: "session-prior",
                        kind: .retrieval,
                        outcome: extra.outcome,
                        promptId: "p-\(seed.id)-1",
                        responseText: nil,
                        masteryBefore: seed.mastery,
                        masteryAfter: seed.mastery,
                        createdAt: now.addingTimeInterval(-extra.daysAgo * LearnerConstants.secondsPerDay)
                    )
                )
            }
            let base = Concept(
                id: seed.id,
                courseId: course.id,
                name: seed.name,
                importance: seed.importance,
                difficulty: seed.difficulty
            )
            concepts.append(
                LearnerModel.applying(
                    LearnerModel.recomputeConceptCache(
                        conceptId: seed.id,
                        difficulty: seed.difficulty,
                        events: events,
                        now: now
                    ),
                    to: base
                )
            )
            _ = index
        }

        let relationships: [ConceptRelationship] = [
            .init(id: "r1", fromId: "c-supply", toId: "c-elasticity", kind: .prerequisite),
            .init(id: "r2", fromId: "c-elasticity", toId: "c-choice", kind: .prerequisite),
            .init(id: "r3", fromId: "c-supply", toId: "c-markets", kind: .prerequisite),
            .init(id: "r4", fromId: "c-fiscal", toId: "c-monetary", kind: .related),
            .init(id: "r5", fromId: "c-supply", toId: "c-fiscal", kind: .related),
        ]

        let prompts: [RetrievalPrompt] = [
            .init(id: "p-c-supply-1", conceptId: "c-supply", question: "What happens to equilibrium price when demand rises and supply is unchanged?", modelAnswer: "Price rises and quantity traded rises until the market clears at the new intersection."),
            .init(id: "p-c-supply-2", conceptId: "c-supply", question: "Why does a binding price ceiling create a shortage?", modelAnswer: "Quantity demanded exceeds quantity supplied at the capped price, so some buyers cannot transact."),
            .init(id: "p-c-elasticity-1", conceptId: "c-elasticity", question: "Why does demand become more elastic when close substitutes are available?", modelAnswer: "Buyers can switch more easily when price rises, so quantity demanded falls by more."),
            .init(id: "p-c-elasticity-2", conceptId: "c-elasticity", question: "If demand is inelastic, what happens to revenue when price increases?", modelAnswer: "Revenue rises because the percentage drop in quantity is smaller than the percentage rise in price."),
            .init(id: "p-c-choice-1", conceptId: "c-choice", question: "What does a budget constraint represent?", modelAnswer: "The combinations of goods a consumer can afford given prices and income."),
            .init(id: "p-c-choice-2", conceptId: "c-choice", question: "At an interior optimum, what is equalized?", modelAnswer: "The marginal utility per dollar (or MRS) equals the price ratio."),
            .init(id: "p-c-markets-1", conceptId: "c-markets", question: "How does perfect competition differ from monopoly in the long run?", modelAnswer: "Competition drives economic profit toward zero; a monopoly can sustain profit with barriers to entry."),
            .init(id: "p-c-markets-2", conceptId: "c-markets", question: "What is the core inefficiency of monopoly pricing?", modelAnswer: "Price above marginal cost, so some mutually beneficial trades do not occur."),
            .init(id: "p-c-fiscal-1", conceptId: "c-fiscal", question: "How can a government spending increase affect aggregate demand?", modelAnswer: "It raises AD directly, and may multiply if households spend the extra income."),
            .init(id: "p-c-fiscal-2", conceptId: "c-fiscal", question: "What is crowding out in a simple closed economy?", modelAnswer: "Higher public borrowing can raise interest rates and reduce private investment."),
            .init(id: "p-c-monetary-1", conceptId: "c-monetary", question: "How does a rate cut typically stimulate spending?", modelAnswer: "Cheaper borrowing and higher asset prices encourage consumption and investment."),
            .init(id: "p-c-monetary-2", conceptId: "c-monetary", question: "Why might monetary policy be weaker at the zero lower bound?", modelAnswer: "Nominal rates cannot fall further, so conventional cuts lose traction."),
            .init(id: "p-c-game-1", conceptId: "c-game", question: "What is a Nash equilibrium?", modelAnswer: "A profile of strategies where no player can gain by unilaterally changing their own strategy."),
            .init(id: "p-c-game-2", conceptId: "c-game", question: "In a prisoner’s dilemma, why is the dominant-strategy outcome inefficient?", modelAnswer: "Each defects to protect themselves, so both do worse than if they could commit to cooperate."),
        ]

        return LearnerSnapshot(
            profile: profile,
            course: course,
            exam: exam,
            concepts: concepts,
            relationships: relationships,
            prompts: prompts,
            events: events
        )
    }
}
